src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,746 @@
+
+    paused_vcpu: Option<Tid>,
+}
+
+fn tid_to_cpuid(tid: Tid) -> usize {
+    tid.get() - 1
+}
+
+/// Converts the inernal index of a vcpu to
+/// the Tid required by GDB
+pub fn cpuid_to_tid(cpu_id: usize) -> Tid {
+    Tid::new(get_raw_tid(cpu_id)).expect("Error translating TID")
+}
+
+/// Converts the inernal index of a vcpu to
+/// the 1 indexed value for GDB
+pub fn get_raw_tid(cpu_id: usize) -> usize {
+    cpu_id + 1
+}
+
+impl FirecrackerTarget {
+    /// Creates a new Target for GDB stub. This is used as the layer between GDB and the VMM it
+    /// will handle requests from GDB and perform the appropriate actions, while also updating GDB
+    /// with the state of the VMM / VCPU's as we hit debug events
+    pub fn new(vmm: Arc<Mutex<Vmm>>, gdb_event: Receiver<usize>, entry_addr: GuestAddress) -> Self {
+        let cpu_count = {
+            vmm.lock()
+                .expect("Exception unlocking vmm")
+                .vcpus_handles
+                .len()
+        };
+
+        let vcpu_state = (0..cpu_count)
+            .map(|_| VCpuState {
+                paused: false,
+                single_step: false,
+            })
+            .collect();
+
+        Self {
+            vmm,
+            entry_addr,
+            gdb_event,
+            // We only support 4 hw breakpoints on x86 this will need to be configurable on arm
+            hw_breakpoints: Vec::with_capacity(4),
+            sw_breakpoints: HashMap::new(),
+            vcpu_state,
+
+            paused_vcpu: Tid::new(1),
+        }
+    }
+
+    fn get_paused_vcpu(&self) -> Result<Tid, Error> {
+        match self.paused_vcpu {
+            Some(res) => Ok(res),
+            None => Err(Error::NoPausedVCpu),
+        }
+    }
+
+    /// This is used to notify the target that the provided Tid
+    /// is in a paused state
+    pub fn notify_paused_vcpu(&mut self, tid: Tid) {
+        let found = &mut self.vcpu_state[tid_to_cpuid(tid)];
+        found.paused = true;
+
+        self.paused_vcpu = Some(tid);
+    }
+
+    fn resume_execution(&mut self) -> Result<(), Error> {
+        let to_resume: Vec<Tid> = self
+            .vcpu_state
+            .iter()
+            .enumerate()
+            .filter_map(|(cpu_id, state)| match state.paused {
+                true => Some(cpuid_to_tid(cpu_id)),
+                false => None,
+            })
+            .collect();
+
+        for tid in to_resume {
+            self.update_kvm_debug(tid)?;
+            self.request_resume(tid)?;
+        }
+
+        self.vcpu_state.iter_mut().for_each(|state| {
+            state.paused = false;
+        });
+
+        self.paused_vcpu = None;
+
+        Ok(())
+    }
+
+    fn reset_vcpu_state(cpu_state: &mut VCpuState) {
+        cpu_state.single_step = false;
+    }
+
+    fn reset_all_states(&mut self) {
+        for value in self.vcpu_state.iter_mut() {
+            Self::reset_vcpu_state(value);
+        }
+    }
+
+    /// This method is used to shutdown the VMM
```

Comment:
Let's have the comments use an active voice, e.g. 
```suggestion
    /// Shuts down the VMM
```

Reply:
Updated them now to use active voice